Расмотрим пример на датасете из репозитория UCI

Описание данных - https://archive.ics.uci.edu/ml/datasets/banknote+authentication#

In [1]:
import pandas as pd
import numpy as np

In [2]:
from google.colab import drive
drive.mount('gdrive')

KeyboardInterrupt: ignored

In [26]:
data = pd.read_csv('/content/citrus.csv')
data.head(5)

name  diameter  weight  red  green  blue
0  orange      2.96   86.76  172     85     2
1  orange      3.91   88.05  166     78     3
2  orange      4.42   95.17  156     81     2
3  orange      4.47   95.60  163     81     4
4  orange      4.48   95.76  161     72     9

Взял достаточно простой датасет, который можно найти тут https://www.kaggle.com/joshmcadams/oranges-vs-grapefruit

In [4]:
data.shape

(10000, 6)

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   name      10000 non-null  object 
 1   diameter  10000 non-null  float64
 2   weight    10000 non-null  float64
 3   red       10000 non-null  int64  
 4   green     10000 non-null  int64  
 5   blue      10000 non-null  int64  
dtypes: float64(2), int64(3), object(1)
memory usage: 468.9+ KB


In [6]:
data.describe()

diameter        weight           red         green          blue
count  10000.000000  10000.000000  10000.000000  10000.000000  10000.000000
mean       9.975685    175.050792    153.847800     76.010600     11.363200
std        1.947844     29.212119     10.432954     11.708433      9.061275
min        2.960000     86.760000    115.000000     31.000000      2.000000
25%        8.460000    152.220000    147.000000     68.000000      2.000000
50%        9.980000    174.985000    154.000000     76.000000     10.000000
75%       11.480000    197.722500    161.000000     84.000000     17.000000
max       16.450000    261.510000    192.000000    116.000000     56.000000

Разбиваем выборку на тренировочную и тестовую части и обучаем модель (в примере - градиентный бустинг)

In [27]:
data.loc[data['name'] == 'orange', 'name'] = 1
data.loc[data['name'] == 'grapefruit', 'name'] = 0

In [8]:
data

name  diameter  weight  red  green  blue
0       1      2.96   86.76  172     85     2
1       1      3.91   88.05  166     78     3
2       1      4.42   95.17  156     81     2
3       1      4.47   95.60  163     81     4
4       1      4.48   95.76  161     72     9
...   ...       ...     ...  ...    ...   ...
9995    0     15.35  253.89  149     77    20
9996    0     15.41  254.67  148     68     7
9997    0     15.59  256.50  168     82    20
9998    0     15.92  260.14  142     72    11
9999    0     16.45  261.51  152     74     2

[10000 rows x 6 columns]

In [35]:
from sklearn.model_selection import train_test_split

x_data = data.drop(['name'], 1)
y_data = data.name

x_train, x_test, y_train, y_test = train_test_split(x_data, y_data, test_size=0.2, random_state=7)

In [10]:
y_data.value_counts()

1    5000
0    5000
Name: name, dtype: int64

In [24]:
pip install catboost

     |████████████████████████████████| 67.3MB 57kB/s 


In [31]:
y_train

4989    1
4498    1
8876    0
670     1
9552    0
       ..
919     1
4307    1
5699    0
537     1
9412    0
Name: name, Length: 8000, dtype: object

In [36]:
import xgboost as xgb

model = xgb.XGBClassifier()

model.fit(x_train, y_train)
y_predict = model.predict(x_test)

In [38]:
import xgboost as xgb
from catboost import CatBoostClassifier

model = CatBoostClassifier()

model.fit(x_train, y_train)
y_predict = model.predict(x_test)

TypeError: ignored

In [19]:
y_test = y_test.to_numpy()

In [34]:
print(type(y_test))
print(type(y_predict))

<class 'pandas.core.series.Series'>
<class 'numpy.ndarray'>


Проверяем качество

In [37]:
from sklearn.metrics import recall_score, precision_score, roc_auc_score, accuracy_score, f1_score

def evaluate_results(y_test, y_predict):
    print('Classification results:')
    f1 = f1_score(y_test, y_predict)
    print("f1: %.2f%%" % (f1 * 100.0)) 
    roc = roc_auc_score(y_test, y_predict)
    print("roc: %.2f%%" % (roc * 100.0)) 
    rec = recall_score(y_test, y_predict, average='binary')
    print("recall: %.2f%%" % (rec * 100.0)) 
    prc = precision_score(y_test, y_predict, average='binary')
    print("precision: %.2f%%" % (prc * 100.0)) 

    
evaluate_results(y_test, y_predict)

Classification results:


ValueError: ignored

## Не понимаю с чем связана эта ошибка. Как ни пытался исправить - не работает. Подскажите пожалуйста, с чем она связана?

### Теперь очередь за PU learning

Представим, что нам неизвестны негативы и часть позитивов

In [ ]:
mod_data = data.copy()
#get the indices of the positives samples
pos_ind = np.where(mod_data.iloc[:,-1].values == 1)[0]
#shuffle them
np.random.shuffle(pos_ind)
# leave just 25% of the positives marked
pos_sample_len = int(np.ceil(0.25 * len(pos_ind)))
print(f'Using {pos_sample_len}/{len(pos_ind)} as positives and unlabeling the rest')
pos_sample = pos_ind[:pos_sample_len]

Using 153/610 as positives and unlabeling the rest


Создаем столбец для новой целевой переменной, где у нас два класса - P (1) и U (-1)

In [ ]:
mod_data['class_test'] = -1
mod_data.loc[pos_sample,'class_test'] = 1
print('target variable:\n', mod_data.iloc[:,-1].value_counts())

target variable:
 -1    1219
 1     153
Name: class_test, dtype: int64


* We now have just 153 positive samples labeled as 1 in the 'class_test' col while the rest is unlabeled as -1.

* Recall that col 4 still holds the actual label

In [ ]:
mod_data.head(10)

0       1        2        3  4  class_test
0  3.62160  8.6661 -2.80730 -0.44699  0          -1
1  4.54590  8.1674 -2.45860 -1.46210  0          -1
2  3.86600 -2.6383  1.92420  0.10645  0          -1
3  3.45660  9.5228 -4.01120 -3.59440  0          -1
4  0.32924 -4.4552  4.57180 -0.98880  0          -1
5  4.36840  9.6718 -3.96060 -3.16250  0          -1
6  3.59120  3.0129  0.72888  0.56421  0          -1
7  2.09220 -6.8100  8.46360 -0.60216  0          -1
8  3.20320  5.7588 -0.75345 -0.61251  0          -1
9  1.53560  9.1772 -2.27180 -0.73535  0          -1

Remember that this data frame (x_data) includes the former target variable that we keep here just to compare the results

[:-2] is the original class label for positive and negative data [:-1] is the new class for positive and unlabeled data

In [ ]:
x_data = mod_data.iloc[:,:-2].values # just the X 
y_labeled = mod_data.iloc[:,-1].values # new class (just the P & U)
y_positive = mod_data.iloc[:,-2].values # original class

### 1. random negative sampling

In [ ]:
mod_data = mod_data.sample(frac=1)
neg_sample = mod_data[mod_data['class_test']==-1][:len(mod_data[mod_data['class_test']==1])]
sample_test = mod_data[mod_data['class_test']==-1][len(mod_data[mod_data['class_test']==1]):]
pos_sample = mod_data[mod_data['class_test']==1]
print(neg_sample.shape, pos_sample.shape)
sample_train = pd.concat([neg_sample, pos_sample]).sample(frac=1)

(153, 6) (153, 6)


In [ ]:
model = xgb.XGBClassifier()

model.fit(sample_train.iloc[:,:-2].values, 
          sample_train.iloc[:,-2].values)
y_predict = model.predict(sample_test.iloc[:,:-2].values)
evaluate_results(sample_test.iloc[:,-2].values, y_predict)

Classification results:
f1: 97.73%
roc: 98.51%
recall: 99.23%
precision: 96.28%


### Домашнее задание

1. взять любой набор данных для бинарной классификации (можно скачать один из модельных с https://archive.ics.uci.edu/ml/datasets.php)
3. сделать feature engineering
4. обучить любой классификатор (какой вам нравится)
5. далее разделить ваш набор данных на два множества: P (positives) и U (unlabeled). Причем брать нужно не все положительные (класс 1) примеры, а только лишь часть
6. применить random negative sampling для построения классификатора в новых условиях
7. сравнить качество с решением из пункта 4 (построить отчет - таблицу метрик)
8. поэкспериментировать с долей P на шаге 5 (как будет меняться качество модели при уменьшении/увеличении размера P)

<b>Бонусный вопрос:</b>

Как вы думаете, какой из методов на практике является более предпочтительным: random negative sampling или 2-step approach?

Ваш ответ здесь: